aspnetcore/security/authentication/configure-oidc-web-authentication.md

```diff
@@ -0,0 +1,317 @@
+using Microsoft.AspNetCore.Authentication.Cookies;
+using Microsoft.AspNetCore.Authentication.OpenIdConnect;
+using Microsoft.IdentityModel.Protocols.OpenIdConnect;
+using Microsoft.IdentityModel.Tokens;
+```
+
+### Setup the configuration properties
+
+Add the OpenID Connect client settings to the application configuration properties. The settings must match the client configuration in the OpenID Connect server. No secrets are persisted in the application settings, the secrets are stored in a Key Vault in production environments or in user secrets in a development environment.
+
+```json
+"OpenIDConnectSettings": {
+  // OpenID Connect URL
+  "Authority": "https://localhost:44318",
+  // client ID from the OpenID Connect server
+  "ClientId": "oidc-pkce-confidential",
+  //"ClientSecret": "--stored-in-user-secrets-or-key-vault--"
+},
+```
+
+### Update the ASP.NET Core pipeline method in the program class.
+
+The UseRouting must be implemented before the UseAuthorization method.
+
+```csharp
+app.UseHttpsRedirection();
+app.UseStaticFiles();
+
+app.UseRouting();
+// Authorization is applied for middleware after the UseAuthorization method
+app.UseAuthorization();
+app.MapRazorPages();
+```
+
+### Force authorization
+
+Add the Authorize attribute to the protected razor pages, for example the Index.cshtml.cs file
+
+```csharp
+[Authorize]
+```
+
+A better way would be to force the whole application to be authorized and opt out for unsecure pages
+
+```csharp
+builder.Services.AddAuthorizationBuilder()
+    .SetFallbackPolicy(new AuthorizationPolicyBuilder()
+    .RequireAuthenticatedUser()
+    .Build());
+```
+
+### Add a new Logout.cshtml and SignedOut.cshtml Razor page to the project
+
+A logout is required to sign-out both the cookie session and the OpenID Connect session. The whole application needs to redirect to the OpenID Connect server to sign-out. After a successful sign-out, the application will open the RedirectUri route.
+
+Implement a default sign-out page and change the Logout razor page code with this:
+
+```csharp
+[Authorize]
+public class LogoutModel : PageModel
+{
+    public IActionResult OnGetAsync()
+    {
+        return SignOut(new AuthenticationProperties
+        {
+            RedirectUri = "/SignedOut"
+        },
+        CookieAuthenticationDefaults.AuthenticationScheme,
+        OpenIdConnectDefaults.AuthenticationScheme);
+    }
+}
+```
+
+The `SignedOut.cshtml` requires the AllowAnonymous attribute.
+
+```csharp
+[AllowAnonymous]
+public class SignedOutModel : PageModel
+{
+    public void OnGet()
+    {
+    }
+}
+```
+
+### Add a login, logout button for the user.
+
+```
+@if (Context.User.Identity!.IsAuthenticated)
+{
+	<li class="nav-item">
+		<a class="nav-link text-dark" asp-area="" asp-page="/Logout">Logout</a>
+	</li>
+
+	<span class="nav-link text-dark">Hi @Context.User.Identity.Name</span>
+}
+else
+{
+	<li class="nav-item">
+		<a class="nav-link text-dark" asp-area="" asp-page="/Index">Login</a>
```

Comment:
Do we think it's worth showing how to initiate a challenge explicitly using `return Challenge()` or `HttpContext.ChallengeAsync()` similar to what we have for signout?

Reply:
fixed, added a Login page using the HttpContext.ChallengeAsync()